Perform basic lower bound constraint analysis

CicadaCinema opened the pull request.

CicadaCinema committed:
bugfix: when computing summary, look at file path and extension before trying to get resolved library

CicadaCinema committed:
add usages command and basic functionality to perform lower bound constraint analysis

CicadaCinema committed:
do not expose 'usages' as a command, this will only be used internally

CicadaCinema committed:
add documentation

CicadaCinema committed:
refactor analysis with a proper reportIssues method

CicadaCinema committed:
use more specific visitor visitMethodInvocation

CicadaCinema committed:
flatten visitMethodInvocation method

CicadaCinema committed:
determine lowest allowed version with no assumptions

CicadaCinema committed:
add name and version fields to PackageShape

CicadaCinema committed:
properly set lowestVersion, currentVersion and constraint when creating LowerBoundConstraintIssue object

CicadaCinema committed:
create helper function getInstalledVersion and clean up

CicadaCinema committed:
refactor analysis functions to accept AnalysisSessions rather than AnalysisContextCollections
also do not download dependencies directly, instead create pointer packages with a single dependency

CicadaCinema committed:
clean up

CicadaCinema committed:
add first test for lower bound constraint analysis

CicadaCinema committed:
check multiple classes for target method

CicadaCinema committed:
fix typos

CicadaCinema committed:
rename PackageAnalysisSession to PackageAnalysisContext

CicadaCinema committed:
rename instances of PackageAnalysisContext
also introduce utility methods for resourceProvider operations

CicadaCinema committed:
finalise lower bound constraint tests to run entirely from yaml

CicadaCinema committed:
test for deep directory structure, also expand functions_methods.yaml test

CicadaCinema committed:
rename/consolidate _PackageAnalysisContext to PackageAnalysisContextWithStderr, making it internal instead of private
also improve test comments/structure

CicadaCinema committed:
fix error in test

jonasfj commented on lib/src/package_analysis/shapes_ext.dart:
```diff
@@ -60,3 +60,19 @@ extension LibraryShapeExt on LibraryShape {
+  /// Returns a list of the names of all the top-level functions which are defined in this package.
+  List<String> get getFunctions {
+    return functions.map((function) => function.name).toList();
+  }
```
Don't use `get...` names..
```suggestion
  List<String> get functionNames {
    return functions.map((function) => function.name).toList();
  }
```

jonasfj commented on lib/src/package_analysis/shapes_ext.dart:
```diff
@@ -60,3 +60,19 @@ extension LibraryShapeExt on LibraryShape {
+  }
+
+  /// Returns a list of the names of all the class methods which are defined in this package.
+  List<String> get getMethods {
```
```suggestion
  List<String> get allMethodNames {
```

But consider doing `bool hasSomeClassWithMethod(String name)`

jonasfj commented on lib/src/package_analysis/summary.dart:
```diff
@@ -11,7 +12,18 @@ Future<PackageShape> summarizePackage(
   PackageAnalysisContext packageAnalysisContext,
```
Just a suggestion to make your life easier :D
```suggestion
  PackageAnalysisContext context,
```

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+      analysisSession.resourceProvider.getFile(path);
+
+  resource.Folder getFolder(String path) =>
+      analysisSession.resourceProvider.getFolder(path);
```
```suggestion
  resource.File file(String path) =>
      analysisSession.resourceProvider.getFile(path);

  resource.Folder folder(String path) =>
      analysisSession.resourceProvider.getFolder(path);
```

jonasfj replied:
Not important, just a suggestion..

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+  resource.Folder getFolder(String path) =>
+      analysisSession.resourceProvider.getFolder(path);
+
+  String? uriToPath(Uri uri) => analysisSession.uriConverter.uriToPath(uri);
```
What kind of URIs to what kind of paths, can we document this?

Also elaborate on what cases it might return `null`.

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+    },
+    'dependencies': {},
+  };
+  pointerPubspec['dependencies'][name] = version;
```
Just a suggestion:
```suggestion
    'dependencies': {
      name: version,
    },
  };
```

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+
+  // construct pubspec for the pointer package
+  final pointerPubspec = <String, dynamic>{
+    'name': 'dummy_package',
```
Consider using dot in the package name:
```suggestion
    'name': 'dummy.package',
```

That way it can't be published, so it can't conflict with a package that exists on pub.dev :D

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+///
+/// A pointer package will be created at [destination] with a single dependency,
+/// pinning the dependency [name] to version [version].
+Future<void> fetchPackageWithPointer({
```
I might call it something else... like `createDummyPackageWithDependency` or `fetchUsingDummyPackage`

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+) async {
+  final dependencyUri = Uri.parse('package:$dependencyName/');
+  final dependencyFilePath = packageAnalysisContext.uriToPath(dependencyUri);
+  print('path for $dependencyName in $packageLocation is $dependencyFilePath');
```
Maybe skip this print line :D

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+  final dependencyUri = Uri.parse('package:$dependencyName/');
+  final dependencyFilePath = packageAnalysisContext.uriToPath(dependencyUri);
+  print('path for $dependencyName in $packageLocation is $dependencyFilePath');
+  return dependencyFilePath == null ? null : path.dirname(dependencyFilePath);
```
Why the `dirname`? Maybe leave a comment explaining that it's because you go a level up to move out of `lib/`, right?

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+/// Ensure that the dependencies of the target package are fetched.
+Future<String?> getDependencyDirectory(
+  PackageAnalysisContext packageAnalysisContext,
+  String packageLocation,
```
We don't need `packageLocation` as an argument
```suggestion
```

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+/// Given the location of a package and the name of one of its dependencies,
+/// return the path of the dependency, or null if it cannot be resolved.
+/// Ensure that the dependencies of the target package are fetched.
+Future<String?> getDependencyDirectory(
```
```suggestion
Future<String?> findPackageDirectory(
```

Or maybe just make an extension method on `packageAnalysisContext`? or even a method on `packageAnalysisContext` called `packageAnalysisContext.packageDirectory(String packageName)` or even just `packageAnalysisContext.packageFolder(String packageName)`.

Maybe, I also suggest that we make it non-nullable?
If the package doesn't exist then something is wrong, right? Like we failed to run `dart pub get` or we have a bug in our analysis code because we are looking for a package that wasn't a dependency or transitive dependency?
So it should never return null, correct? unless there is a bug in our code.

Thus, I suggest that we simply make it non-nullable and if we fail to resolve it (in cases where `dependencyFilePath == null`), we just throw  an `StateError` saying that the referenced package doesn't exist.
Because it's always an error if we try to analyze a package that we don't depend on..

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+
+/// Fetch all the hosted dependencies used by a package at a given location
+Future<Map<String, HostedDependency>> getHostedDependencies(
+    String targetPackageDir) async {
```
Just a suggestion, feel free to ignore.
```suggestion
Future<Map<String, HostedDependency>> getHostedDependencies(
  String targetPackageDir,
) async {
```

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+  final allDependencies = Pubspec.parseYaml(
+          await File(path.join(targetPackageDir, 'pubspec.yaml'))
+              .readAsString())
+      .dependencies;
```
Just an idea... Because FormatException and IOException should never happen here, if they do then it's because we gave a path that's not really a package.
Handling them and throwing a generic StateError is nice because we don't want someone calling  `getHostedDependencies` to handle `FormatException` or `IOException`, someone calling `getHostedDependencies` should always ensure they are pointing to a real package.

```suggestion
try {
  final allDependencies = Pubspec.parseYaml(
          await File(path.join(targetPackageDir, 'pubspec.yaml'))
              .readAsString())
      .dependencies;
} on FormatException {
  throw StateError('Failed to parse pubspec.yaml from "$targetPackageDir" check that it is a valid package');
} on IOException {
  throw StateError('Failed to read pubspec.yaml from "$targetPackageDir" check that it is a valid package');
}
```

CicadaCinema replied:
I don't think this error handling is necessary

Pubspec.parseYaml() is invoked in three places so far - in two of them the relevant pubspec file is generated beforehand, and in the last its presence is verified (before the summary is generated):
```dart
if (!await File(path.join(packageLocation, 'pubspec.yaml')).exists()) {
  throw StateError('The target directory must contain a package.');
}
```

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+      constraint.includeMin &&
+      versions.contains(constraint.min!)) {
+    return constraint.min!;
+  }
```
Do we need this special case? Isn't it fine to just always do the thing below.
I bet that this isn't slow, so keeping the code simple is probably better than optimizing for a fast-path (which might not be that fast).

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
+      return version;
+    }
+  }
+  return null;
```
You might not even need a method for doing this :D
```suggestion
  return versions.firstWhere(constraint.allows, orElse: () => null);
```

jonasfj commented on test/package_analysis/common.dart:
```diff
@@ -0,0 +1,54 @@
+import 'package:analyzer/file_system/physical_file_system.dart';
+import 'package:path/path.dart' as path;
+
+String getConfigTemplate({
```
```suggestion
String _createPackageConfig({
```

jonasfj commented on test/package_analysis/common.dart:
```diff
@@ -0,0 +1,54 @@
+      'generatorVersion': '2.17.3'
+    });
+
+String getPubspecTemplate({
+  required String packageName,
```
```suggestion
String _createPubspec({
  required String packageName,
```

jonasfj commented on test/package_analysis/common.dart:
```diff
@@ -0,0 +1,54 @@
+  required String packageName,
+  required String packageVersion,
+}) =>
+    'name: $packageName\nversion: $packageVersion';
```
```suggestion
    'name: $packageName\nversion: $packageVersion\n';
```

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,241 @@
+    required this.targetDependencies,
+    required this.dependencyInstalledVersions,
+  });
+
```
```suggestion
  /// The package URI for the currently analyzed unit, typically on the form: `package:foo/foo.dart`.
  ///
  /// Defaults to a dummy value, to ensure it's never `null`.
  Uri currentFile = Uri.parse('package:$rootPackage/'); // we might have to set this in the constructor.

  @override
  void visitCompilationUnit(ResolvedUnitResult unit) {
    currentFile = unit.uri;
    //TODO: Investigate if we have to use: packageAnalysisContext.analysisSession.uriConverter.pathToUri(unit.path);
    
    super.visitCompilationUnit(unit);
    
    // Restore dummy value
    currentFile = Uri.parse('package:$rootPackage/');
  }
```

jonasfj replied:
Now you have the path to the file being analyzed, so you can use the filename in the `SourceSpan`...
and you create the contents as necessary for creating source spans..

CicadaCinema replied:
done:
```dart
/// The package URI for the currently analyzed unit, typically on the form
/// `package:package_name/library_name.dart`.
late Uri currentUri;
  ```

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,241 @@
+
+    final element = node.methodName.staticElement;
+
+    if (element == null) {
+      warning(
+          'Element associated with identifier ${node.methodName.name} could not be resolved.');
+      return;
+    }
```
```suggestion
    if (element == null) {
      // We cannot statically resolve what is invoked when a method is called on a variable with
      // dynamic type. In this case we just do nothing, since we can't know what is being called.
      return;
    }
```

jonasfj replied:
No need to print a warning, since this is an expect case that can naturally happen.

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,241 @@
+    // then this symbol cannot be analyzed
+    if (packageName == null ||
+        packageName == rootPackage ||
+        packageName == 'flutter') {
```
There are other dependencies SDK dependencies than just `flutter´.
Maybe we should explicitly ignore all packages that are not `HostedDependencies`.

We don't care about git-dependencies and path-dependencies. Those won't exist on pub.dev, but if someone were to analyze their local application then they might have such dependencies.

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,241 @@
+          'Failed to resolve subclass of enclosingElement ${enclosingElement.toString()}.');
+    }
+
+    switch (constraintIssue) {
```
Using a switch here is a bit messy, hehe :D

But the logic seems okay..

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,241 @@
+
+    final dependencyShape = dependencySummaries[packageName]!;
+    final enclosingElement = element.enclosingElement;
+    bool? constraintIssue;
```
```suggestion
    bool constraintIssue = false;
```

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,241 @@
+/// Given a target package and its dependencies (at their lower bound),
+/// analyze the target package and return a List of any found issues - where a
+/// symbol usage cannot be found in the relevant dependency's PackageShape
+Future<List<LowerBoundConstraintIssue>> reportIssues({
```
```suggestion
Future<List<LowerBoundConstraintIssue>> lowerBoundConstraintAnalysis({
```

CicadaCinema commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,204 @@
 
-  AnalysisContext contextFor(String path) =>
-      analysisContextCollection.contextFor(path);
+abstract class PackageAnalysisContext {
```
add field `packageName` - root package name, for error messages
add field `dependencies` with type Map<String,Version> that maps from (hosted) dependency name to installed version
also add field `pubspec` and field `packagePath`

CicadaCinema committed:
add static fields to ClassShape:
staticGetters
staticSetters
staticMethods

CicadaCinema committed:
sort the fields of PackageShape.classes in normalizePackageShape

CicadaCinema committed:
remove unnecessary String packageLocation argument from two helper functions

CicadaCinema committed:
address nits

CicadaCinema committed:
expand PackageAnalysisContext with more fields, reducing the number of necessary arguments of lowerBoundConstraintAnalysis and summarizePackage

CicadaCinema committed:
add SourceSpan references to LowerBoundConstraintIssue instances
also add util for downloading json with package metadata from pub,
as well as other bug fixes

CicadaCinema committed:
do not analyse package/dependency if download failed

CicadaCinema committed:
major bugfix: analyse the lowest version of each dependency rather than the current version

CicadaCinema committed:
add span to list of references if an issue is encountered for the first time

CicadaCinema committed:
account for static methods in lower bound constraint analysis

CicadaCinema committed:
remove unnecessary libraryPath argument

CicadaCinema committed:
ignore generics

CicadaCinema committed:
use dummy package when downloading target to perform lower bound constraint analysis

CicadaCinema committed:
add test using inheritance

CicadaCinema committed:
add visitor to be able to analyze properties, hopefully without breaking too much
common checks shared between visitors are in processIdentifier, which throws AnalysisException in case further analysis is not required/possible

CicadaCinema committed:
expand test to ensure class members which are unchanged between versions do not raise issues

CicadaCinema committed:
add one more small dataset of packages to analyse

CicadaCinema committed:
do not crash analysis if a dependency is not found in the json doc

CicadaCinema committed:
do not filter out generics before setting enclosingElement

CicadaCinema committed:
detect parentElement more accurately

CicadaCinema committed:
do not create duplicate elements in ClassShape fields when adding class members from superclasses

CicadaCinema committed:
add test for prior commit

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,235 @@
+
+  /// The name of the singular dependency of this target dummy package, or null
+  /// if the package being analysed does not use a dummy.
+  final String? targetPackageName;
```
How is this different from `packageName`, do we ever need to know `packageName` after the initial setup (if these two are different?)

CicadaCinema replied:
addressed in https://github.com/CicadaCinema/pana/pull/2/commits/7e09995c47283da2c271325567dc918abe7f00c8

jonasfj commented on lib/src/package_analysis/common.dart:
```diff
@@ -1,12 +1,235 @@
+  late final Pubspec pubspec =
+      Pubspec.parseYaml(readFile(path.join(packagePath, 'pubspec.yaml')));
+
+  /// A [Map] storing metadata about the hosted dependencies of the package
+  /// being analyzed.
+  late final Map<String, HostedDependency> dependencies = Map.fromEntries(
+      pubspec
+          .dependencies.entries
+          .where((dependency) => dependency.value is HostedDependency)
+          .map(
+              (entry) => MapEntry(entry.key, entry.value as HostedDependency)));
+
+  /// A [Map] storing metadata about the hosted dependencies of the package
+  /// being analyzed, given that the package associated with this
+  /// [PackageAnalysisContext] is a dummy package with a single non-transitive
+  /// dependency.
+  late final Map<String, HostedDependency> targetDependencies = Map.fromEntries(
```
Why do we need both?

CicadaCinema replied:
addressed in https://github.com/CicadaCinema/pana/pull/2/commits/7e09995c47283da2c271325567dc918abe7f00c8

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,378 @@
+  late Element parentElement;
+
+  /// The [SourceSpan] corresponding to the invocation currently being analyzed.
+  late SourceSpan span;
```
Don't it make more sense to make this a `FileSpan` for the current file (and rename it `sourceFile`).

```dart
sourceFile = SourceFile.fromString(
  context.readFile(context.uriToPath(currentUri)!),
  url: currentUri,
)
```

Then later you just do:
```dart
issues[elementId]?.references.add(sourceFile.span(identifier.offset, identifier.end));
```

jonasfj replied:
Ah, this is populated by `processIdentifier`, in that case you might want to document it.. this is slightly confusing. Another option is to return values from `processIdentifier`.

jonasfj commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,378 @@
+    dependencyShape = dependencySummaries[packageName]!;
+  }
+
+  /// Populate the various properties of this visitor based on information from
+  /// this symbol, throwing an [AnalysisException] if the symbol does not need
+  /// to be/cannot be analyzed.
```
We could also return some data class that holds the different properties, that might be less confusing.

CicadaCinema committed:
encapsulate methods like context.contains* in extension

CicadaCinema committed:
eliminate false positives by requiring issues to no longer exist in the latest version of the dependency
also tweak summarizePackage to accept a packagePath instead of the name of the package

CicadaCinema committed:
correct findMinAllowedVersion to not require sorted list of versions as input,
also make the function more readable

CicadaCinema committed:
remove unnecessary getters

CicadaCinema committed:
convert getDependencyDirectory function to method of PackageAnalysisContext

CicadaCinema committed:
merge packageName and targetPackageName, packagePath and targetPackagePath, dependencies and targetDependencies

CicadaCinema committed:
make getInstalledVersion a method of PackageAnalysisContext

CicadaCinema committed:
remove accidental debug print

CicadaCinema committed:
add performance monitoring and more verbose warnings for found issues

CicadaCinema committed:
add placeholder for allPackages

CicadaCinema committed:
do not crash if getAllFiles is passed a resource.Folder which does not exist

CicadaCinema committed:
handle case m.hashCode where m is Function() by returning early

CicadaCinema committed:
perform input validation, add new test

CicadaCinema committed:
eliminate ambiguity between a pointer package and a dummy package

CicadaCinema committed:
eliminate checkArgs function as it is only used once

CicadaCinema committed:
replace instances of dummy_package and test_package with dummy.package and test.package (only where they refer to package names)

CicadaCinema committed:
if we have decided not to analyse a package, still delete its temporary directory

CicadaCinema committed:
add test for exporting an external symbol

CicadaCinema committed:
sort class members more nicely

CicadaCinema committed:
simplify switch statement

CicadaCinema committed:
implement typedefs in summary and account for typedef aliases in lower bound constraint analysis to filter out false positives
also make false positive checks clearer

CicadaCinema committed:
remove unused method
also remove accidental print

CicadaCinema committed:
remove unused function

CicadaCinema committed:
make yaml tests more concise

CicadaCinema committed:
make lbcanalysis command single-purpose - it now only analyses a single named package
now no analysis operation relies on the large metadata JSON file

CicadaCinema committed:
correct small issue with method renaming

CicadaCinema committed:
add batchlbca command to analyse multiple packages in parallel

CicadaCinema commented on lib/src/package_analysis/package_analysis.dart:
```diff
@@ -214,6 +216,153 @@ class LowerBoundConstraintAnalysisCommand extends Command {
+                'Analysis of package $targetPackageName completed with a non-zero exit code.');
+          }
+        } catch (e) {
+          stderr.write(
```
Does this only execute if the analysis command times out? (if the analysis crashes then `process` should end with a non-zero exit code, right?)

CicadaCinema replied:
yes, now we do `} on TimeoutException {`

CicadaCinema commented on lib/src/package_analysis/package_analysis.dart:
```diff
@@ -214,6 +216,153 @@ class LowerBoundConstraintAnalysisCommand extends Command {
+                'Analysis of package $targetPackageName completed with a non-zero exit code.');
+          }
+        } catch (e) {
+          stderr.write(
```
How should we handle deletion of the temporary folder(s) created by the analysis command if it times out? Maybe it's okay that they can't be handled if the analysis crashes _unexpectedly_... Should the analysis command have its own timeout so it can gracefully deal with packages like `google_fonts` which take a long time to analyse?

CicadaCinema replied:
wait for https://github.com/dart-lang/sdk/issues/49861 to be fixed, then delete the temp directory within the child process on receiving a sigterm, like this cross-platform approach https://github.com/dart-lang/webdev/pull/370/files#diff-75e84e99a12c4a689b4e5e202ad664388b38c852a465edff8fd8bbe384c18a11R59

CicadaCinema committed:
do not write empty log files

CicadaCinema committed:
remove util which generated large JSON metadata file, we have no use for it any more

CicadaCinema committed:
replace calls to write() with writeln()

CicadaCinema committed:
add extensions to the summary, write an associated test

CicadaCinema committed:
change LibraryShape.replace() method to LibraryShape.replaceAll() to reflect its usage

CicadaCinema committed:
migrate to newer version of the analyzer
also add package:pool as a direct dependency

CicadaCinema committed:
detect missing extension members while performing lower bound constraint analysis

CicadaCinema committed:
make caching optional when fetching sorted package versions

CicadaCinema committed:
allow setting a user-defined pub hosted url when fetching target package using a dummy

CicadaCinema committed:
generalize function lowerBoundConstraintAnalysis to reduce code duplication between cli and tests

CicadaCinema committed:
remember to normalize extensions too

CicadaCinema committed:
make stderr less noisy when the lowest allowed version of a dependency cannot be fetched

CicadaCinema committed:
expand issue detection messages with readable issue metadata

CicadaCinema committed:
add capability to analyse all packages, not just those returned by name completion endpoint

CicadaCinema committed:
flush and close any files being written to

CicadaCinema committed:
fall back on allPackages.length when provided package count to analyse is too high

CicadaCinema committed:
make astVisitor final, and use proper syntax for serving packages

CicadaCinema committed:
reduce packageCountToAnalyze more gracefully when it is too high

CicadaCinema committed:
calculate packageCountToAnalyze even more gracefully

CicadaCinema committed:
simplify behaviour of fetchUsingDummyPackage

CicadaCinema committed:
make language clearer around the current dependency version check and the typedef check

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+}
+
+// TODO: rename this class? it is still useful to have this structure even if we are not sure whether this is a real issue or a false positive - consider LowerBoundConstraintIssueResult or similar
+class LowerBoundConstraintIssue {
```
rename this class? it is still useful to have this structure even if we are not sure whether this is a _real_ issue or a _false positive_ - consider `LowerBoundConstraintIssueResult` or similar

CicadaCinema replied:
PotentialLowerBoundConstraintIssue

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+  /// The [ElementKind] of the [Element] corresponding to the missing
+  /// [identifier], one of [ElementKind.FUNCTION], [ElementKind.METHOD],
+  /// [ElementKind.GETTER], [ElementKind.SETTER].
+  final ElementKind kind;
```
Use enumeration somehow to assert that this member is one of several supported values

CicadaCinema replied:
create my own `ElementKind`s and enums

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+  /// The [ElementKind] of the parent [Element] of the missing [identifier],
+  /// [parentName], one of [ElementKind.CLASS], [ElementKind.EXTENSION], or null
+  /// if [identifier] is not a member of a class/extension.
+  final ElementKind? parentKind;
```
Use enumeration somehow to assert that this member is one of several supported values

CicadaCinema replied:
create my own `ElementKind`s and enums

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+    required PackageShape package,
+    String? classNameAlias,
+  }) {
+    switch (kind) {
```
any way to make this cleaner? we need to switch/branch depending on the value of `kind` as well as `parentKind`

CicadaCinema replied:
do this by nesting if statements

CicadaCinema replied:
did it with nested switches since both members were enums

CicadaCinema commented on lib/src/package_analysis/package_analysis.dart:
```diff
@@ -21,50 +29,348 @@ class SummaryCommand extends Command {
+        late final String dependencyRepository;
+        // TODO: not sure if nested try-finally here is OK
+        // I'm doing this because usually there are no found issues and no need to create an http client
+        try {
```
not sure if nested try-finally here is OK
I'm doing this because usually there are no found issues and no need to create an http client

CicadaCinema replied:
close client in outer finally

CicadaCinema commented on test/package_analysis/package_summary_test.dart:
```diff
@@ -1,80 +1,67 @@
-      final physicalProvider = PhysicalResourceProvider();
-      final provider = OverlayResourceProvider(physicalProvider);
+      // TODO: can we rely on this path being empty and our testing not conflicting with physical files on disk?
+      final packagePath = path.canonicalize(path.join('test_package'));
```
can we rely on this path being empty and our testing not conflicting with physical files on disk?
we do not need to create a unique temporary directory _on disk_ because in this case the tests are run on an in-memory test package
no IO disk operations take place during the test
it is enough to know that we have a path which does not coincide with an existing directory

CicadaCinema replied:
OK

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+      // we cannot statically resolve what was invoked
+      return;
+    }
+    element = node.methodName.staticElement!;
```
does `element.declaration` help here?
what kind of syntax does it cover?

CicadaCinema replied:
https://pub.dev/documentation/analyzer/latest/dart_element_element/Element/declaration.html

CicadaCinema replied:
OK - `node.methodName.staticElement!` already gets us to the required `Element`

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+  // accesses to properties of objects can also be represented as
+  // PrefixedIdentifier nodes in cases where the target is also a simple identifier.
+  @override
+  void visitPrefixedIdentifier(PrefixedIdentifier node) {
```
do we need visitPropertyAccess too?
https://pub.dev/documentation/analyzer/latest/dart_ast_ast/PropertyAccess-class.html
> The access of a property of an object.
> Note, however, that accesses to properties of objects can also be represented as [PrefixedIdentifier](https://pub.dev/documentation/analyzer/latest/dart_ast_ast/PrefixedIdentifier-class.html) nodes in cases where the target is also a simple identifier.

What is [Expression](https://pub.dev/documentation/analyzer/latest/dart_ast_ast/Expression-class.html) and should we consider cases where it is the `node.target`?

CicadaCinema replied:
test:
```
class Foo {
  int a = 5;
}
Foo f() => Foo();

f().a; // property access ??

final t = f();
t.a; // prefixed identifier

```

CicadaCinema committed:
migrate TODOs to a GitHub review

CicadaCinema committed:
create pubspec file before writing to it

CicadaCinema committed:
create temporary directory and assign the pub cache to this directory during tests
this fixes flaky test behaviour where a globalPackageServer port is re-used across tests

CicadaCinema commented on lib/src/package_analysis/lower_bound_constraint_analysis.dart:
```diff
@@ -0,0 +1,630 @@
+    );
+  } on ProcessException catch (exception) {
+    throw AnalysisException(
+        'Failed to download target package $targetName with error code ${exception.errorCode}: ${exception.message}');
```
Because we are downloading packages using a dummy, we have no package metadata other than the name of the package when executing `lbcanalysis` - when the SDK constraints are too low, `pub get` fails here.

Maybe we can query an endpoint like https://pub.dev/api/packages/retry BEFORE spawning the `lbcanalysis` process, and only spawn this process if two conditions are met:
1) the lower bound sdk constraint is >= 2.12.0
2) the current sdk version satisfies the sdk constraint

otherwise, do not launch `lbcanalysis` and skip onto the next package

although I'm not sure whether making ~30,000 http requests over the course of ~8 hours (between 1 and 2 requests per second) is OK, with no cache in place. though making one GET request is certainly faster than: launching analysis, writing the pubspec to disk and seeing whether `pub get` succeeds/fails

CicadaCinema replied:
OK to pre-compute this before starting any analysis operations

CicadaCinema committed:
remove consuming fields of class _LowerBoundConstraintVisitor,
instead making identifyDependencyName and processIdentifier return data classes with metadata on the identifier being analysed

CicadaCinema committed:
avoid chance of conflict with physical files on disk

CicadaCinema committed:
rename LowerBoundConstraintIssue to PotentialLowerBoundConstraintIssue

CicadaCinema commented:
TODO: add license headers

CicadaCinema commented:
TODO: rename package analysis to (public) api analysis and move package_analysis.dart to bin/api_analysis.dart

CicadaCinema commented:
TODO: can https://pub.dev/documentation/analyzer/latest/dart_ast_ast/MethodInvocation/realTarget.html help with handling cascade notation? replace target with realTarget and write accompanying test.

DONE

CicadaCinema committed:
do not attempt to analyse packages not supporting null safety, or those whose sdk constraint is not satisfied by the current version of the dart sdk

CicadaCinema committed:
support PropertyAccess nodes

CicadaCinema committed:
adjust test syntax.yaml

CicadaCinema committed:
enable detection of tear-offs

jonasfj commented on lib/src/package_analysis/package_analysis.dart:
```diff
@@ -21,50 +30,384 @@ class SummaryCommand extends Command {
+
+          // wait for the analysis to complete or time out after 10 minutes
+          final exitCode =
+              await process.exitCode.timeout(const Duration(minutes: 10));
```
This can throw a `TimeoutException`

jonasfj commented on lib/src/package_analysis/package_analysis.dart:
```diff
@@ -21,50 +30,384 @@ class SummaryCommand extends Command {
+            stderr.writeln(
+                'Analysis of package $targetPackageName completed with a non-zero exit code.');
+          }
+        } catch (e) {
```
You might want to do `} on TimeoutException {` and handle this separately, because then you can display a message specific to timeout issues..

Also, catching all exceptions/errors can hide internal error that you then fail to discover.

CicadaCinema committed:
canonicalize path before setting overlay

CicadaCinema committed:
fix bug where private accessors were being displayed in summary

CicadaCinema committed:
add enum summary test

CicadaCinema committed:
slim down lower_bound_constraint_analysis.dart

CicadaCinema committed:
use own enums for PotentialLowerBoundConstraintIssue.kind and PotentialLowerBoundConstraintIssue.parentKind,
fix a bug where enums were not being considered and add a test

CicadaCinema committed:
remove unneeded assertion - the same StateError would just appear later on

CicadaCinema committed:
catch TimeoutException instead of catching everything

CicadaCinema committed:
add null check

CicadaCinema committed:
add another null check

CicadaCinema committed:
fix regression due to previously added null checks

CicadaCinema committed:
handle cascade notation

CicadaCinema commented on lib/src/package_analysis/kind.dart:
```diff
@@ -0,0 +1,35 @@
+
+enum Kind { method, function, getter, setter }
+
+enum ParentKind { Class, Extension, Enum }
```
OK to capitalise the values in this enum? ignore lint constant_identifier_names? `class`,  `extension` and `enum` are already keywords. maybe `classKind` (or `classK` for brevity) instead? ideally the notation for `Kind` and `ParentKind` should be identical.

sigurdm replied:
I would go with `classKind` etc.

CicadaCinema commented on lib/src/package_analysis/shapes.dart:
```diff
@@ -153,6 +214,23 @@ class GlobalPropertyShape extends GlobalShapeBase {
+@sealed
+@JsonSerializable()
+class TypedefShape extends GlobalShapeBase {
+  /// The [ClassShape.id] that this typedef points to, or null if defined as function type
```
TODO: test to see if https://dart-lang.github.io/linter/lints/comment_references.html no longer triggers here (and elsewhere in this file) now that https://github.com/Dart-Code/Dart-Code/issues/4114 is fixed

CicadaCinema replied:
### before
```
C:\Users\Atom\Desktop\pana>dart --version
Dart SDK version: 2.17.6 (stable) (Tue Jul 12 12:54:37 2022 +0200) on "windows_x64"

C:\Users\Atom\Desktop\pana>dart analyze .\lib\src\package_analysis
Analyzing package_analysis...

   info • shapes.dart:69:18 • Only reference in scope identifiers in doc comments. • comment_references
   info • shapes.dart:73:18 • Only reference in scope identifiers in doc comments. • comment_references
   info • shapes.dart:77:18 • Only reference in scope identifiers in doc comments. • comment_references
   info • shapes.dart:81:18 • Only reference in scope identifiers in doc comments. • comment_references
   info • shapes.dart:85:18 • Only reference in scope identifiers in doc comments. • comment_references
   info • shapes.dart:89:18 • Only reference in scope identifiers in doc comments. • comment_references
   info • shapes.dart:242:12 • Only reference in scope identifiers in doc comments. • comment_references

7 issues found.
```

### after
```
C:\Users\Atom\Desktop\pana>dart --version
Dart SDK version: 2.19.0-146.0.dev (dev) (Fri Aug 26 18:55:25 2022 -0700) on "windows_x64"

C:\Users\Atom\Desktop\pana>dart analyze .\lib\src\package_analysis
Analyzing package_analysis...
No issues found!
```

CicadaCinema committed:
add constructors to the summary

CicadaCinema committed:
add support for some basic annotations on classes and extensions
(but not their members, yet)

CicadaCinema committed:
rename values in enum ParentKind

CicadaCinema committed:
correct class names in docstrings
